boost_histogram/cpp/tag.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import absolute_import, division, print_function
+
+del absolute_import, division, print_function
+
+import numpy as _np
+
+
+class Locator(object):
+    __slots__ = "offset"
+
+    def __init__(self, offset=0):
+        if not isinstance(offset, int):
+            raise ValueError("The offset must be an integer")
+
+        self.offset = offset
+
+    def __add__(self, offset):
+        from copy import copy
+
+        other = copy(self)
+        other.offset += offset
+        return other
+
+    def __sub__(self, offset):
+        from copy import copy
+
+        other = copy(self)
+        other.offset -= offset
+        return other
+
+    def _print_self_(self):
+        return ""
+
+    def __repr__(self):
+        s = self.__class__.__name__
+        s += self._print_self_()
+        if self.offset != 0:
+            s += " + " if self.offset > 0 else " - "
+            s += str(abs(self.offset))
+        return s
+
+
+class loc(Locator):
+    __slots__ = ("value",)
+
+    def __init__(self, value, offset=0):
+        super(loc, self).__init__(offset)
+        self.value = value
+
+    def _print_self_(self):
+        return "({0})".format(self.value)
+
+    def __call__(self, axis):
+        return axis.index(self.value) + self.offset
+
+
+class underflow(Locator):
+    __slots__ = ()
+
+    def __call__(self, axis):
+        return -1 + self.offset
+
+
+underflow = underflow()
+
+
+class overflow(Locator):
+    __slots__ = ()
+
+    def __call__(self, axis):
+        return len(axis) + self.offset
+
+
+overflow = overflow()
+
+
+class at(object):
```

Comment:
I see, at has the cpp behavior and loc has the Pythonic behavior, nice.

Reply:
The system is quite flexible and powerful; users can make all sorts of things - defining at yourself is trivial; I'm using classes here just to get a nicer repr. Once we support Python projects/rebins (0.7.0 or so), we can do the same for the step argument items as well, that will finish the rest of the original UHI proposal.

Note that PEP 8 does allow these to be lowercase; classes that mostly used like functions or values don't have to be CamelCase.